relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+		return false, errors.Annotate(err, "read binlog header")
+	}
+
+	if !bytes.Equal(buff, replication.BinLogFileHeader) {
+		return false, errors.Errorf("binlog file %s header not valid, got % X, expect % X", filename, buff, replication.BinLogFileHeader)
+	}
+	return true, nil
+}
+
+// checkFormatDescriptionEventExist checks if the file has a valid FormatDescriptionEvent.
+func checkFormatDescriptionEventExist(filename string) (bool, error) {
+	// FormatDescriptionEvent always follows the binlog file header
+	exist, err := checkBinlogHeaderExist(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		return false, errors.Errorf("no binlog file header at the beginning for %s", filename)
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// check whether only the file header
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	fs, err := f.Stat()
+	if err != nil {
+		return false, errors.Errorf("get stat for %s", filename)
+	} else if fs.Size() == int64(fileHeaderLen) {
+		return false, nil // only the file header
+	}
+
+	// seek to the beginning of the FormatDescriptionEvent
+	_, err = f.Seek(int64(fileHeaderLen), io.SeekStart)
+	if err != nil {
+		return false, errors.Annotatef(err, "seek to %d for %s", fileHeaderLen, filename)
+	}
+
+	// parse a FormatDescriptionEvent
+	var found bool
+	onEventFunc := func(e *replication.BinlogEvent) error {
+		if e.Header.EventType != replication.FORMAT_DESCRIPTION_EVENT {
+			return errors.Errorf("got %+v, expect FormatDescriptionEvent", e.Header)
+		}
+		found = true
+		return nil
+	}
+
+	eof, err := replication.NewBinlogParser().ParseSingleEvent(f, onEventFunc)
+	if err != nil {
+		return false, errors.Annotatef(err, "parse %s", filename)
+	} else if eof {
+		return false, errors.Annotatef(io.EOF, "parse %s", filename)
+	}
+	return found, nil
+}
+
+// checkIsDuplicateEvent checks if the event is a duplicate event in the file.
+// NOTE: handle cases when file size > 4GB
+func checkIsDuplicateEvent(filename string, ev *replication.BinlogEvent) (bool, error) {
+	// 1. check event start/end pos with the file size
+	fs, err := os.Stat(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "get stat for %s", filename)
+	}
+	evStartPos := int64(ev.Header.LogPos - ev.Header.EventSize)
+	evEndPos := int64(ev.Header.LogPos)
+	if fs.Size() <= evStartPos {
+		return false, nil // the event not in the file
+	} else if fs.Size() < evEndPos {
+		// the file can not hold the whole event, often because the file is corrupt
+		return false, errors.Errorf("file size %d is between event's start pos (%d) and end pos (%d)",
+			fs.Size(), evStartPos, evEndPos)
+	}
+
+	// 2. use a FileReader to read events from the start pos of the passed-in event
+	rCfg := &reader.FileReaderConfig{
+		EnableRawMode: true,
+	}
+	r := reader.NewFileReader(rCfg)
```

Comment:
can we read file directly if we know event size and raw data of `ev`?

Reply:
we can do the read by using the file directly, but we also need to decode it to a binlog event that needs the _FORMAT_DESCRIPTION_EVENT_, so use `FileReader.StartSyncByPos` is simpler.

Reply:
Must need to decode?

Reply:
changed to compare the raw data.